### DISCUSSION TITLE
Welcome to posbe Discussions!

### DISCUSSION THREAD

**Discussion by aasumitro** [aasumitro/posbe#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by williamluisan** [comment]
Just saw the architecture in details, epic bro !

So each module has it's own service and repository.
Actually what is *cfg and *en means in the diagram? (me still noob xD)
And I am not really clear with the "repository" concept per module, can pass me a reference? need to read about it.

Related to the diagram, are the modules able to interact with each other? if so, is gRPC method used for it?

@aasumitro

**Reply by aasumitro** [reply]
1. cfg stands for config (from /pkg/cfg), and en stands for the engine (gin.Engine)
2. arch  & concept:
- https://medium.easyread.co/golang-clean-archithecture-efd6d7c43047
- https://medium.easyread.co/trying-clean-architecture-on-golang-2-44d615bf8fdf
- https://www.geeksforgeeks.org/single-responsibility-in-solid-design-principle/
- https://github.com/mcholismalik/solid-in-golang
- https://s8sg.medium.com/solid-principle-in-go-e1a624290346
- https://manakuro.medium.com/clean-architecture-with-go-bce409427d31
- https://itnext.io/golang-and-clean-architecture-19ae9aae5683
3. depending on your needs
@williamluisan

**Reply by williamluisan** [reply]
I see, I also think *cfg is config too at first, but not sure tought it was another term. But *en is really not clear for me (just a noob thing xD).
Okay bro, thanks for the references, will take a look a bit.

**Reply by aasumitro** [reply]
The diagram is not completed yet xD

**Reply by aasumitro** [reply]
WDYT @williamluisan 🤔🤔🤔
![](https://www.plantuml.com/plantuml/svg/vPNFIiD048VlUOezAcrUe9HGYXJ4Iz7UoxGxZGERsJ0xUoYjxsx_R1h581YXAkOooNpvNRxTdPq41T5OkGnPIRTcwYBMGgp1x3AJdW7PaPbIUSzgnkFvGr4SIecC0egsA2u_vdcROQXmP9vKYCksQp6iXMebpn-XRkj313bMDu918Ggm3iqAP6ZMiC6XyVgCSjicreAWQ9UW8tC7YWkgiAMnjfzGYmuxwKjsAesHICuIbxs435QDp-MxoEmSkT3mVObNL39ptajF7ybcjuu4_vQDH--hVZ8sHu-T9axaQng2WiRm8mOMi7BulnNCXLXHVh64A7HLCs-lalq-p_UxTfIUUVfgS1ehHvPU4pYTi8v-sdSTe5cDR89iO-LU0G00)

**Reply by aasumitro** [reply]
multi buildings, multi floors

**Reply by williamluisan** [reply]
Is this the full database structure?
so neat.

Questions:
- What is "Printers" in this context, "Printers" means the tool for paper printing?
- "Tables" is per "Floor" or per "Room", if per "Floor" why?

**Reply by aasumitro** [reply]
- nope, still considered
- Yup, Thermal Printer (e.g: cashier bill receipt, order item list to bar/kitchen)
- yea rooms and tables per floor (tables for restaurant) (rooms for karaoke maybe? or private resto) rooms and tables for dine-in (restaurant) or reservation (karaoke, etc)
🤔

current schemas:
<img width="411" alt="Screenshot 2022-12-06 at 21 57 50" src="https://user-images.githubusercontent.com/8925311/205931594-acae18e4-0a83-4cc3-8673-a7fdd6baff6a.png">

**Reply by williamluisan** [reply]
so nice bro.. maybe you missed, I think the "status" column is important for each `tables` and `rooms` table, did not see it in current schema.

**Reply by aasumitro** [reply]
yes, the status for each table is still considered. really hard to think about splitting up this project from monolith to ms with db per service in the future.

imo, table and room status can be stored on cache repo (like redis) and can be handled with some service (named `notify_svc` maybe) that impl sse (server-sent event) for real-time push/broadcast

notify_svc will stream data from a message broker (like rabbitmq) and push to the front end


```
[transaction_service]:  a new transaction/transaction update with table/room
--> [create data to sql]
--> [pub data to mq]

[store (room/table) service] 
<-- [sub data from mq]
--> [update cache data]
--> [pub data change]

[notify_svc] 
<-- [sub data chage]
--> send change to fe
```

wdyt?

**Reply by williamluisan** [reply]
Agreed, this is a very good idea bro, and probably move to ms in the future would be good also.
Never try this kind of data flow method on real project, but got the concept and it's very good, and if this concept applied, ms approach is the most suitable for it I think.

Then, that's mean the "status" column term here is to represent the current real-time status of the entity right? I tought before it was only to store the non real-time status like "active", "inactive", "enabled" or "disabled". Instead, it will represents status of the entity like "used", "taken", "closed", etc.

Am I correctly understand with the "status" column term?

**Reply by aasumitro** [reply]
yea thats correct. -> **it will represent the status of the entity** 

active[and/or]inactive will define in the feature flag, ref: https://github.com/aasumitro/posbe/blob/cf60f58a19d6ec7ea313cb586c661036c72ea7ab/db/migrations/20221202035850_add_store_data.up.sql#L25

**Reply by williamluisan** [reply]
Sasadikile mo coba lia depe dalam2 ini project, banyak skali ilmu baru noh 😆 😆 👍🏾 !
Unit testing so ada Mit?

**Reply by aasumitro** [reply]
unit testing yang mana? rata2 itu sudah unit + integration testing, tiap ada file baru, usecase baru, repo baru pasti dibuatkan unit testing + integration test. kecuali kalau nda salah yang redis untuk fungsi CacheFirst data karna baru test tadi mocking dpe redis server wkwwkwkkw dan baru impl di role handler

liat ada yang ada _test.go di masing2 folder itu testingnya
pokoknya kalau interface atau fungsi sdh di pake pasti udah ada ut atau it nya, digambar yang ijo2 itu dia
<img width="2056" alt="Screenshot 2022-12-10 at 00 52 45" src="https://user-images.githubusercontent.com/8925311/206752095-fd3f78d8-ed5f-4a1f-885f-95ec2d32a832.png">

**Reply by williamluisan** [reply]
Mantap ini, spann 👍🏾 👍🏾 !
